app/src/main/java/com/cmput301f20t21/bookfriends/ui/library/OwnedViewModel.java

```diff
@@ -1,101 +1,104 @@
     private final BookRepository bookRepository = BookRepository.getInstance();
-    
-    private MutableLiveData<List<Book>> books;
-    private MutableLiveData<Integer> updatedPosition;
-    private MutableLiveData<BOOK_ERROR> errorMessageObserver;
+
+    private MutableLiveData<List<Book>> books = new MutableLiveData<>(new ArrayList<>());
+    private List<Book> bookData = books.getValue();
+
+    private MutableLiveData<Integer> updatedPosition = new MutableLiveData<>(0);
+    private MutableLiveData<BOOK_ERROR> errorMessageObserver = new MutableLiveData<>();
+
+    public OwnedViewModel() {
+        fetchBooks();
+    }
 
     public LiveData<List<Book>> getBooks() {
-        if (books == null) {
-            books = new MutableLiveData<>();
-            fetchBooks();
-        }
         return books;
     }
 
     public LiveData<Integer> getUpdatedPosition() {
-        if (updatedPosition == null) {
-            updatedPosition = new MutableLiveData<>(0);
-        }
         return updatedPosition;
     }
 
     public LiveData<BOOK_ERROR> getErrorMessageObserver() {
-        if (errorMessageObserver == null) {
-            errorMessageObserver = new MutableLiveData<>();
-        }
         return errorMessageObserver;
     }
 
-    public void deleteBook(Book book, OnSuccessCallbackWithMessage<Book> successCallback, OnFailCallback failCallback) {
-        bookRepository.delete(book.getId()).addOnCompleteListener(
-                deleteBookTask -> {
-                    if (deleteBookTask.isSuccessful()) {
-                        Uri imageUri = book.getImageUri();
-                        if (imageUri != null) {
-                            bookRepository.deleteImage(book.getId()).addOnCompleteListener(
-                                    deleteImageTask -> successCallback.run(book)
-                            );
-                        } else {
-                            successCallback.run(book);
-                        }
-                    } else {
-                        failCallback.run();
-                    }
-                }
-        );
+    public Book getBookByIndex(Integer index) {
+        return bookData.get(index);
+    }
+
+    public void addBook(Book book) {
+        bookData.add(book);
+        books.setValue(bookData);
+    }
+
+    public void updateBook(Book oldBook, Book updatedBook) {
+        Integer indexToUpdate = bookData.indexOf(oldBook);
+        if (indexToUpdate != -1) {
+            bookData.set(indexToUpdate, updatedBook);
+            books.setValue(bookData);
+        } else {
+            errorMessageObserver.setValue(BOOK_ERROR.FAIL_TO_EDIT_BOOK);
+        }
+    }
+
+    public void deleteBook(Book book) {
+        bookRepository.delete(book.getId())
+                .addOnSuccessListener(
+                        res -> {
+                            if (book.getImageUri() != null) {
+                                bookRepository.deleteImage(book.getId());
```

Comment:
we don't care if image delete success or not if the book is deleted from its own collection.